File: scripts/validate_workflows.py
#!/usr/bin/env python3
"""Validate importable n8n workflow JSON files.

The n8n export format stores workflow connections by node *name*, not by node
ID. This validator follows that convention and catches the failure modes that
make imported templates feel broken: missing names, dangling connections,
duplicate nodes, thin canvases, and missing setup notes.
"""

import json
import sys
from pathlib import Path
from typing import Optional


MIN_NODE_COUNT = 6
MIN_STICKY_NOTE_COUNT = 2


def discover_workflows(repo_root: Path):
    """Return numbered workflow exports such as 01-example/workflow.json."""

    return sorted(repo_root.glob("[0-9][0-9]-*/workflow.json"))


def _load_json(workflow_path: Path) -> tuple[Optional[dict], list[str]]:
    try:
        with workflow_path.open() as workflow_file:
            workflow = json.load(workflow_file)
    except json.JSONDecodeError as exc:
        return None, [f"JSON parse error: {exc}"]
    except FileNotFoundError:
        return None, [f"File not found: {workflow_path}"]

    if not isinstance(workflow, dict):
        return None, ["Workflow must be a JSON object"]

    return workflow, []


def validate_workflow(workflow_path: Path) -> list[str]:
    """Validate a single workflow.json file."""

    errors = []
    workflow, load_errors = _load_json(workflow_path)
    if load_errors:
        return load_errors

    assert workflow is not None

    for field in ("name", "nodes", "connections"):
        if field not in workflow:
            errors.append(f"Missing required field: {field}")

    if errors:
        return errors

    nodes = workflow.get("nodes", [])
    connections = workflow.get("connections", {})

    if not isinstance(nodes, list):
        errors.append("Field 'nodes' must be a list")
        nodes = []
    elif len(nodes) < MIN_NODE_COUNT:
        errors.append(f"Expected at least {MIN_NODE_COUNT} nodes, got {len(nodes)}")

    if not isinstance(connections, dict):
        errors.append("Field 'connections' must be an object")
        connections = {}

    node_names = set()
    node_ids = set()
    duplicate_names = set()
    duplicate_ids = set()

    for index, node in enumerate(nodes):
        if not isinstance(node, dict):
            errors.append(f"Node at index {index} must be an object")
            continue

        node_name = node.get("name")
        node_id = node.get("id")

        if not isinstance(node_name, str) or not node_name.strip():
            errors.append(f"Node at index {index} is missing a name")
        elif node_name in node_names:
            duplicate_names.add(node_name)
        else:
            node_names.add(node_name)

        if node_id is not None:
            if not isinstance(node_id, str) or not node_id.strip():
                errors.append(f"Node '{node_name or index}' has an invalid id")
            elif node_id in node_ids:
                duplicate_ids.add(node_id)
            else:
                node_ids.add(node_id)

    for name in sorted(duplicate_names):
        errors.append(f"Duplicate node name: {name}")

    for node_id in sorted(duplicate_ids):
        errors.append(f"Duplicate node id: {node_id}")

    sticky_notes = [node for node in nodes if node.get("type") == "n8n-nodes-base.stickyNote"]
    if len(sticky_notes) < MIN_STICKY_NOTE_COUNT:
        errors.append(f"Expected at least {MIN_STICKY_NOTE_COUNT} sticky notes, got {len(sticky_notes)}")

    for source_name, source_connections in connections.items():
        if source_name not in node_names:
            errors.append(f"Connection references unknown source node: {source_name}")
            continue

        if not isinstance(source_connections, dict):
            errors.append(f"Connections for '{source_name}' must be an object")
            continue

        for branch_index, branch in enumerate(source_connections.get("main", [])):
            if not isinstance(branch, list):
                errors.append(f"Main branch {branch_index} for '{source_name}' must be a list")
                continue

            for connection in branch:
                if not isinstance(connection, dict):
                    errors.append(f"Connection from '{source_name}' must be an object")
                    continue

                target_name = connection.get("node")
                if target_name not in node_names:
                    errors.append(f"Connection from '{source_name}' references unknown target node: {target_name}")

    return errors


def main() -> int:
    """Run validation on all numbered workflows."""

    repo_root = Path(__file__).parent.parent
    workflows = discover_workflows(repo_root)

    if not workflows:
        print("ERROR: No workflows found")
        return 1

    all_passed = True
    for workflow_path in workflows:
        print(f"Validating {workflow_path.parent.name}...", end=" ")
        errors = validate_workflow(workflow_path)

        if errors:
            print("FAILED")
            for error in errors:
                print(f"  - {error}")
            all_passed = False
        else:
            print("PASSED")

    if all_passed:
        print(f"\nAll {len(workflows)} workflows valid.")
        return 0

    print("\nValidation failed for one or more workflows.")
    return 1


if __name__ == "__main__":
    sys.exit(main())

File: scripts/validate-workflows.ts
import { mkdir, readdir, readFile, writeFile } from "node:fs/promises";
import path from "node:path";
import process from "node:process";

type N8nNode = {
  id?: unknown;
  name?: unknown;
  type?: unknown;
};

type N8nConnection = {
  node?: unknown;
};

type N8nWorkflow = {
  name?: unknown;
  nodes?: unknown;
  connections?: unknown;
};

type CatalogEntry = {
  slug: string;
  name: string;
  nodeCount: number;
  stickyNoteCount: number;
  triggerTypes: string[];
  integrationTypes: string[];
  workflowPath: string;
};

type ValidationResult = {
  filePath: string;
  errors: string[];
  catalogEntry?: CatalogEntry;
};

const MIN_NODE_COUNT = 6;
const MIN_STICKY_NOTE_COUNT = 2;
const WORKFLOW_DIR_PATTERN = /^\d{2}-.+$/;
const repoRoot = process.cwd();

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

function asString(value: unknown): string | undefined {
  return typeof value === "string" && value.trim().length > 0 ? value : undefined;
}

function nodeTypeLabel(nodeType: string): string {
  return nodeType.replace(/^n8n-nodes-base\./, "");
}

async function discoverWorkflowFiles(): Promise<string[]> {
  const entries = await readdir(repoRoot, { withFileTypes: true });
  return entries
    .filter((entry) => entry.isDirectory() && WORKFLOW_DIR_PATTERN.test(entry.name))
    .map((entry) => path.join(repoRoot, entry.name, "workflow.json"))
    .sort();
}

async function readWorkflow(filePath: string): Promise<N8nWorkflow> {
  const raw = await readFile(filePath, "utf8");
  return JSON.parse(raw) as N8nWorkflow;
}

function validateWorkflow(filePath: string, workflow: N8nWorkflow): ValidationResult {
  const errors: string[] = [];
  const slug = path.basename(path.dirname(filePath));

  const workflowName = asString(workflow.name);
  if (!workflowName) {
    errors.push("Missing required string field: name");
  }

  if (!Array.isArray(workflow.nodes)) {
    errors.push("Missing required array field: nodes");
  }

  if (!isRecord(workflow.connections)) {
    errors.push("Missing required object field: connections");
  }

  if (errors.length > 0) {
    return { filePath, errors };
  }

  const nodes = workflow.nodes as N8nNode[];
  const connections = workflow.connections as Record<string, unknown>;
  const nodeNames = new Set<string>();
  const nodeIds = new Set<string>();
  const nodeTypes = new Set<string>();
  const triggerTypes = new Set<string>();
  const duplicateNames = new Set<string>();
  const duplicateIds = new Set<string>();

  if (nodes.length < MIN_NODE_COUNT) {
    errors.push(`Expected at least ${MIN_NODE_COUNT} nodes, got ${nodes.length}`);
  }

  for (const [index, node] of nodes.entries()) {
    if (!isRecord(node)) {
      errors.push(`Node at index ${index} must be an object`);
      continue;
    }

    const nodeName = asString(node.name);
    const nodeId = asString(node.id);
    const nodeType = asString(node.type);

    if (!nodeName) {
      errors.push(`Node at index ${index} is missing a name`);
    } else if (nodeNames.has(nodeName)) {
      duplicateNames.add(nodeName);
    } else {
      nodeNames.add(nodeName);
    }

    if (node.id !== undefined) {
      if (!nodeId) {
        errors.push(`Node '${nodeName ?? index}' has an invalid id`);
      } else if (nodeIds.has(nodeId)) {
        duplicateIds.add(nodeId);
      } else {
        nodeIds.add(nodeId);
      }
    }

    if (nodeType) {
      nodeTypes.add(nodeType);
      if (nodeType.includes("Trigger") || nodeType.endsWith(".webhook")) {
        triggerTypes.add(nodeTypeLabel(nodeType));
      }
    }
  }

  for (const name of [...duplicateNames].sort()) {
    errors.push(`Duplicate node name: ${name}`);
  }

  for (const nodeId of [...duplicateIds].sort()) {
    errors.push(`Duplicate node id: ${nodeId}`);
  }

  const stickyNoteCount = nodes.filter((node) => node.type === "n8n-nodes-base.stickyNote").length;
  if (stickyNoteCount < MIN_STICKY_NOTE_COUNT) {
    errors.push(`Expected at least ${MIN_STICKY_NOTE_COUNT} sticky notes, got ${stickyNoteCount}`);
  }

  for (const [sourceName, sourceConnections] of Object.entries(connections)) {
    if (!nodeNames.has(sourceName)) {
      errors.push(`Connection references unknown source node: ${sourceName}`);
      continue;
    }

    if (!isRecord(sourceConnections)) {
      errors.push(`Connections for '${sourceName}' must be an object`);
      continue;
    }

    const mainConnections = sourceConnections.main;
    if (mainConnections === undefined) {
      continue;
    }

    if (!Array.isArray(mainConnections)) {
      errors.push(`Main connections for '${sourceName}' must be an array`);
      continue;
    }

    for (const branch of mainConnections) {
      if (!Array.isArray(branch)) {
        errors.push(`Connection branch for '${sourceName}' must be an array`);
        continue;
      }

      for (const connection of branch as N8nConnection[]) {
        if (!isRecord(connection)) {
          errors.push(`Connection from '${sourceName}' must be an object`);
          continue;
        }

        const targetName = asString(connection.node);
        if (!targetName || !nodeNames.has(targetName)) {
          errors.push(`Connection from '${sourceName}' references unknown target node: ${targetName}`);
        }
      }
    }
  }

  const integrationTypes = [...nodeTypes]
    .filter((nodeType) => nodeType !== "n8n-nodes-base.stickyNote")
    .map(nodeTypeLabel)
    .sort();

  return {
    filePath,
    errors,
    catalogEntry: {
      slug,
      name: workflowName ?? slug,
      nodeCount: nodes.length,
      stickyNoteCount,
      triggerTypes: [...triggerTypes].sort(),
      integrationTypes,
      workflowPath: path.relative(repoRoot, filePath),
    },
  };
}

async function writeCatalog(entries: CatalogEntry[]): Promise<void> {
  const catalogPath = path.join(repoRoot, "catalog", "workflows.json");
  await mkdir(path.dirname(catalogPath), { recursive: true });
  await writeFile(catalogPath, `${JSON.stringify(entries, null, 2)}\n`);
}

async function main(): Promise<void> {
  const shouldWriteCatalog = process.argv.includes("--write-catalog");
  const workflowFiles = await discoverWorkflowFiles();

  if (workflowFiles.length === 0) {
    console.error("ERROR: No numbered workflow.json files found.");
    process.exitCode = 1;
    return;
  }

  const results: ValidationResult[] = [];
  for (const filePath of workflowFiles) {
    try {
      const workflow = await readWorkflow(filePath);
      results.push(validateWorkflow(filePath, workflow));
    } catch (error) {
      const message = error instanceof Error ? error.message : String(error);
      results.push({ filePath, errors: [`Unable to read workflow JSON: ${message}`] });
    }
  }

  const failed = results.filter((result) => result.errors.length > 0);
  for (const result of results) {
    const slug = path.basename(path.dirname(result.filePath));
    if (result.errors.length > 0) {
      console.error(`FAIL ${slug}`);
      for (const error of result.errors) {
        console.error(`  - ${error}`);
      }
    } else {
      console.log(`PASS ${slug}`);
    }
  }

  if (failed.length > 0) {
    console.error(`\n${failed.length} workflow(s) failed validation.`);
    process.exitCode = 1;
    return;
  }

  const catalogEntries = results
    .map((result) => result.catalogEntry)
    .filter((entry): entry is CatalogEntry => entry !== undefined);

  if (shouldWriteCatalog) {
    await writeCatalog(catalogEntries);
    console.log(`\nCatalog written with ${catalogEntries.length} workflows.`);
  } else {
    console.log(`\nAll ${catalogEntries.length} workflows valid.`);
  }
}

main().catch((error: unknown) => {
  const message = error instanceof Error ? error.message : String(error);
  console.error(message);
  process.exitCode = 1;
});
